packages/website/src/components/editor/createProvideTwoslashInlay.ts

```diff
@@ -0,0 +1,100 @@
+// The following code is adapted from the code in microsoft/TypeScript-Website.
+// Source: https://github.com/microsoft/TypeScript-Website/blob/c8b2ea8c8fc216c163fe293650b2666c8563a67d/packages/playground/src/twoslashInlays.ts
+// License: https://github.com/microsoft/TypeScript-Website/blob/c8b2ea8c8fc216c163fe293650b2666c8563a67d/LICENSE-CODE
+
+import type Monaco from 'monaco-editor';
+import type * as ts from 'typescript';
+
+import type { SandboxInstance } from './useSandboxServices';
+
+const twoslashQueryRegex = /^(\s*\/\/\s*\^\?)\s*$/gm;
+function findTwoshashQueries(code: string): RegExpExecArray[] {
+  let match: RegExpExecArray | null = null;
+  const matches: RegExpExecArray[] = [];
+  while ((match = twoslashQueryRegex.exec(code))) {
+    matches.push(match);
+  }
+  return matches;
+}
+
+export function createTwoslashInlayProvider(
+  sandbox: SandboxInstance,
+): Monaco.languages.InlayHintsProvider {
+  return {
+    provideInlayHints: async (
+      model,
+      _,
+      cancel,
+    ): Promise<Monaco.languages.InlayHintList> => {
+      const worker = await sandbox.getWorkerProcess();
+      if (model.isDisposed() || cancel.isCancellationRequested) {
+        return {
+          hints: [],
+          dispose(): void {
+            /* nop */
+          },
+        };
+      }
+
+      const queryMatches = findTwoshashQueries(model.getValue());
+
+      const results: Monaco.languages.InlayHint[] = [];
+
+      for (const result of await Promise.all(
+        queryMatches.map(q => resolveInlayHint(q)),
+      )) {
+        if (result) {
+          results.push(result);
+        }
+      }
+
+      return {
+        hints: results,
+        dispose(): void {
+          /* nop */
+        },
+      };
+
+      async function resolveInlayHint(
+        queryMatch: RegExpExecArray,
+      ): Promise<Monaco.languages.InlayHint | undefined> {
+        const end = queryMatch.index + queryMatch[1].length - 1;
+        const endPos = model.getPositionAt(end);
+        const inspectionPos = new sandbox.monaco.Position(
+          endPos.lineNumber - 1,
+          endPos.column,
+        );
+        const inspectionOff = model.getOffsetAt(inspectionPos);
+
+        const hint = await (worker.getQuickInfoAtPosition(
+          'file://' + model.uri.path,
+          inspectionOff,
+        ) as Promise<ts.QuickInfo | undefined>);
+        if (!hint?.displayParts) {
+          return;
+        }
+
+        // Make a one-liner
+        let text = hint.displayParts
+          .map(d => d.text)
+          .join('')
+          .replace(/\r?\n\s*/g, ' ');
+        if (text.length > 120) {
+          text = text.slice(0, 119) + '...';
+        }
+
+        const inlay: Monaco.languages.InlayHint = {
```

Comment:
[Style] Nit: the function already has a return type, so I don't think a separate variable declaration gives us much?

```suggestion
        return {
```